src/v/storage/disk_log_impl.cc

```diff
@@ -798,11 +833,13 @@ ss::future<> disk_log_impl::remove_segment_permanently(
           "Segment has outstanding locks. Might take a while to close:{}",
           s->reader().filename());
     }
-    return s->close()
-      .handle_exception([s](std::exception_ptr e) {
-          vlog(stlog.error, "Cannot close segment: {} - {}", e, s);
-      })
-      .finally([this, s] { _probe.segment_removed(); });
+    return _readers_cache->evict_on_segment_remove(s).then([this, s] {
```

Comment:
same concern as cache::stop. seems like we could invalidate the readers' entry_guard destructor to avoid waiting on the readers' finally future to resolve?

Reply:
we have to wait since close requires a write lock.

Reply:
the point was that it adds unnecessary coupling and code. when the cache is stopped it can arrange for all entry guards to become noops (breaking dependency) and then all of this code where we have to wrap shutdown sequences in the stop continuation of the cache reader would seem to go away.